adin2111: OA support and fixes

spectrum70 opened the pull request:
Started to work on this driver, to optimize and add missing features.

spectrum70 committed:
drivers: ethernet: adin2111: fix imask1 tx_ready_mask

Fix IAMSK1 TX_READY_MASK bitfield position.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 committed:
drivers: ethernet: adin2111: always append crc32 to the eth frame

Enable appending of a crc32 at the end of the frame by the MAC,
always. This is always needed since the driver is not adding it.

This field has nothing to do with Generic SPI protocol-related
8-bit CRC, so this patch removes the CONFIG_ETH_ADIN2111_SPI_CFG0
choice related to this setting.

Testing without this flag set, packets are not forwareded in the
network, since the driver is not appending any crc32 header
to the frame.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

jpmur commented:
LGTM

spectrum70 committed:
drivers: ethernet: adin2111: add Open Alliance SPI support

Add Open Alliance spi protocol support.

Open Alliance is a chunk-based SPI protocol, based on sending
over SPI an ethernet frame divided in smaller chunks, using a
specific 32-bit header for each chunk transferred. All chunks
can be sent or received by a single dma transfer.

Default mode is set to Open Alliance SPI without protection,
since the adin2111 dev. board comes shipped this way.

Tested:
- Open Alliance SPI, no protection (default board shipped)
- Open Alliance SPI, protection
- Generic SPI, no crc
- Generic SPI, with crc8

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 committed:
drivers: ethernet: adin2111: fix generic spi withotut crc8 mode

After some debugging related to non-working "Generic SPI without CRC"
mode in eval_adin2111_ebz (CONFIG_ETH_ADIN2111_SPI_CFG0=n), noticed
that even after proper STATUS0 RESETC bit detection, registers,
for a certain period (some msecs) still reads as zero.

This patch fixes adin2111_await_device and with it the
"Generic SPI without CRC" mode.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 committed:
drivers: ethernet: adin2111: remove unused header_len variable

Non functional change, removing unused variable producing
compilation warning in "Generic SPI without CRC8" mode.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 commented:
Add other 2 small patches.

spectrum70 commented:
Hi,

Add Open Alliance SPI support.

This is actually the most complex patch of the set, btw, i reused existing Microchip Open Alliance module for this.
Waiting for any Maintainer feedbacks, especially on the way i am zeroing gpio structure, to be able to control chip select by sw, since there may be better ways (see @@ -781,7 +916,7)

carlocaione commented on drivers/ethernet/Kconfig.adin2111:
```diff
@@ -49,6 +49,24 @@ config ETH_ADIN2111_SPI_CFG0
+DT_ETH_ADIN2111_PATH := $(dt_nodelabel_path,adin2111)
+DT_ETH_ADIN2111_SPI_OA := $(dt_node_bool_prop,$(DT_ETH_ADIN2111_PATH),spi-oa)
+
+config ETH_ADIN2111_SPI_OA
+	bool "SPI_OA"
+	default $(DT_ETH_ADIN2111_SPI_OA)
+	help
+	  Must be set when Open Alliance SPI is configured by HW bootstrap deep-switches.
```
Why is this needed? Can't we just use `DT_INST_NODE_HAS_PROP` or equivalent in the code without going through one additional Kconfig symbol?

spectrum70 replied:
This was an intermediate choice, since there is the CRC option (CFG0) actually in Kconfig. In this way all options are visible also in menuconfig.:

(2048) Stack size for a thread that processes ADIN IRQ
(2) Priority for internal incoming packet handler
(100) IP buffer timeout
[*] SPI_CFG0
[*] SPI_OA
[ ] SPI_OA_PROTECTION

In a next step i can set all the options in devicetree only, eventually.

carlocaione replied:
This is totally redundant and while there could be overlapping between DT and Kconfig it is alway better to just put the hardware configuration in DT when possible.

Intermediate steps are only going to increase the workload for all the parts involved (you, reviewers, maintaners, etc...) so I'd just suggest to move all the options into the DT already in this PR.

GeorgeCGV replied:
@spectrum70 my mistake was to place the CRC option into Kconfig.

Ideally, the CRC should be moved into the DT. Some adin eval boards were recently added; and there it might be necessary to control CRC on/off at runtime (_due to board config switches_).

spectrum70 replied:
Ok i proceed as required.

carlocaione commented on drivers/ethernet/Kconfig.adin2111:
```diff
@@ -49,6 +49,24 @@ config ETH_ADIN2111_SPI_CFG0
+
+DT_ETH_ADIN2111_SPI_OA_PROTECTION := $(dt_node_bool_prop,$(DT_ETH_ADIN2111_PATH),spi-oa-protection)
+
+config ETH_ADIN2111_SPI_OA_PROTECTION
+	bool "SPI_OA_PROTECTION"
+	default $(DT_ETH_ADIN2111_SPI_OA_PROTECTION)
+	depends on $(DT_ETH_ADIN2111_SPI_OA)
+	help
+	  Must be set when Open Alliance SPI protection is configured by HW bootstrap deep-switches.
```
ditto

carlocaione commented:
> Waiting for any Maintainer feedbacks, especially on the way i am zeroing gpio structure, to be able to control chip select by sw, since there may be better ways (see @@ -781,7 +916,7)

@gmarull ^

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -779,7 +916,7 @@ static int adin2111_await_device(const struct device *dev)
 static int adin2111_init(const struct device *dev)
 {
-	const struct adin2111_config *cfg = dev->config;
+	struct adin2111_config *cfg = (struct adin2111_config *)dev->config;
```
What is the reason for `const` to disappear?

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -882,11 +1052,7 @@ static int adin2111_init(const struct device *dev)
 	}
 
-#if CONFIG_ETH_ADIN2111_SPI_CFG0
 	val |= ADIN2111_CONFIG2_CRC_APPEND;
```
The CRC is now always enabled.

spectrum70 replied:
This is related to eth frame crc32, not connected to spi crc8

GeorgeCGV replied:
Rechecked RM, indeed.

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -758,10 +891,14 @@ static int adin2111_await_device(const struct device *dev)
 	/* await reset complete (RESETC) and clear it */
 	for (count = 0U; count < ADIN2111_RESETC_AWAIT_RETRY_COUNT; ++count) {
-		ret = eth_adin2111_reg_read(dev, ADIN2111_STATUS0, &val);
+		ret = eth_adin2111_reg_read(dev, ADIN2111_PHYID, &val);
```
Not sure about that change. The RM mentions STATUS0 poll. Will need to test.

Do you use HW reset? If no, then it could be related to clock stabilization delay (_done within hw reset_).

spectrum70 replied:
Thaks, 

i tested mainly using the SW reset. Anyway i see that keeping the reset button pressed adin2111 clock stops.

Reading a read-only reg currently may not be the best, sure, but it works reliably, and allows "Generic SPI without CRC" to work.

What about if i move the check in the phy code ?

Anyway, asking also suggestions to the adi team tight now.



I see that

GeorgeCGV replied:
@spectrum70 in cases when hw reset is not used does increasing the delay from 10ms to `ADIN2111_HW_BOOT_DELAY_MS` helps?

GeorgeCGV replied:
> Anyway, i am asking also suggestions to the adi team right now.

Was there any response from them?

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -805,6 +942,25 @@ static int adin2111_init(const struct device *dev)
+		 * OA requires sw chip select control
+		 */
+		ctx->cs_gpio = cfg->spi.config.cs.gpio;
+		cfg->spi.config.cs.gpio = spec;
```
That is a bit odd as driver config is supposed to be immutable.

It would be better to solve during config definition with macros: `COND_CODE_1` and `DT_INST_NODE_HAS_PROP`.

When OA property is present, then `spi.config.cs.gpio` can be set to `{0, 0, 0}`, otherwise `SPI_DT_SPEC_INST_GET` is used.

That will require OA to be a DT property as proposed in https://github.com/zephyrproject-rtos/zephyr/pull/68532/files#r1482973313.

spectrum70 replied:
Thanks for the feedback, will do as suggestd.

GeorgeCGV commented on drivers/ethernet/eth_adin2111_priv.h:
```diff
@@ -92,7 +101,7 @@
 #define ADIN2111_IMASK1_P1_RX_RDY_MASK		BIT(4)
 /*!< Mask Bit for TX_FRM_DONE. Generic SPI only.*/
-#define ADIN2111_IMASK1_TX_RDY_MASK		BIT(4)
+#define ADIN2111_IMASK1_TX_RDY_MASK		BIT(3)
```
Great catch.

GeorgeCGV commented:
@spectrum70 thank you for the addition, cleanup, and fixes. Tested legacy spi with crc, it works. I am ok with a merge as soon as the driver config is kept immutable (_cs pin is set during config definition_).

spectrum70 commented:
In this PR:
- remove CONFIG_ stuff for Open Alliance, used DT_PROP and COND_CODE_1
- add oa and oa_protection to use/call Open Alliance features.

spectrum70 commented:
> @spectrum70 thank you for the addition, cleanup, and fixes. Tested legacy spi with crc, it works. I am ok with a merge as soon as the driver config is kept immutable (_cs pin is set during config definition_).

Thanks to you, a pleasure to work on this driver, nice and clean code.

gmarull commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -781,7 +915,7 @@ static int adin2111_await_device(const struct device *dev)
 static int adin2111_init(const struct device *dev)
 {
-	const struct adin2111_config *cfg = dev->config;
+	struct adin2111_config *cfg = (struct adin2111_config *)dev->config;
```
please, add const back and remove the unnecessary cast

gmarull commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -852,6 +1000,11 @@ static int adin2111_init(const struct device *dev)
 	}
 
+	if (ctx->oa) {
+		/* OA mode, device needs some time to respond */
+		k_msleep(1);
```
"some time" feels too vague here (magic sleep), isn't there a minimum value specified in the datasheet?

gmarull commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -839,6 +982,11 @@ static int adin2111_init(const struct device *dev)
 	}
 
+	if (ctx->oa) {
+		/* OA mode, device needs some time to respond */
+		k_msleep(10);
```
same here, magic value

gmarull commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -966,21 +1133,49 @@ static const struct ethernet_api adin2111_port_api = {
+		.id = dev_id,									\
+		.spi = {									\
+			.bus = DEVICE_DT_GET(DT_BUS(DT_DRV_INST(inst))),			\
+			.config = {								\
+				.frequency = DT_INST_PROP(inst, spi_max_frequency),		\
+				.operation = (ADIN2111_SPI_OPERATION) |				\
+					DT_INST_PROP(inst, duplex) |				\
+					DT_INST_PROP(inst, frame_format) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_cpol),		\
+						    SPI_MODE_CPOL, (0)) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_cpha),		\
+						    SPI_MODE_CPHA, (0)) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_hold_cs),		\
+						    SPI_HOLD_ON_CS, (0)),			\
+				.slave = DT_REG_ADDR(DT_DRV_INST(inst)),			\
+				.cs = {								\
+					.gpio = COND_CODE_1(DT_INST_PROP_OR(inst, spi_oa, 0),	\
+						({0, 0, 0}),					\
+						(SPI_CS_GPIOS_DT_SPEC_GET(DT_DRV_INST(inst)))),	\
+					.delay = 0,						\
+				}								\
```
aren't SPI DT spec macros sufficient here?

gmarull commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -966,21 +1133,49 @@ static const struct ethernet_api adin2111_port_api = {
-		.spi = SPI_DT_SPEC_INST_GET(inst, ADIN2111_SPI_OPERATION, 1),			\
+		.id = dev_id,									\
+		.spi = {									\
+			.bus = DEVICE_DT_GET(DT_BUS(DT_DRV_INST(inst))),			\
```
please don't use DT_DRV_INST() when `DT_INST_*` macros exist

gmarull commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -966,21 +1133,49 @@ static const struct ethernet_api adin2111_port_api = {
 		.reset = GPIO_DT_SPEC_INST_GET_OR(inst, reset_gpios, { 0 }),			\
 	};											\
+	static struct oa_tc6 name##_oa_tc6_##inst = {						\
+		.cps = 64,									\
```
is this used? can't struct be const?

GeorgeCGV replied:
Struct internals are modified within `oa_tc6.c`, it seems to store config and data.

gmarull commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -968,21 +1137,49 @@ static const struct ethernet_api adin2111_port_api = {
+						    SPI_HOLD_ON_CS, (0)),			\
+				.slave = DT_REG_ADDR(DT_DRV_INST(inst)),			\
+				.cs = {								\
+					.gpio = COND_CODE_1(DT_INST_PROP_OR(inst, spi_oa, 0),	\
+						({0, 0, 0}),					\
+						(SPI_CS_GPIOS_DT_SPEC_GET(DT_DRV_INST(inst)))),	\
```
if the whole objective of `spi-oa` is to override CS control, why not enforce it via a build assert, eg

```
BUILD_ASSERT(DT_INST_HAS_PROP(inst, spi_oa) && !DT_PROP_HAS_IDX(DT_INST_BUS(inst), DT_INST_REG_ADDR(inst));
```
then simply initialize SPI with `SPI_DT_SPEC_INST_GET`.

Another option would be to modify SPI API with a flag to override CS control even if non-NULL.

gmarull replied:
ping here

carlocaione commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -966,21 +1137,49 @@ static const struct ethernet_api adin2111_port_api = {
 		.buf = name##_buffer_##inst,							\
+		.oa = DT_INST_PROP_OR(inst, spi_oa, 0),						\
+		.tc6 = &name##_oa_tc6_##inst,							\
+		.cs_gpio = SPI_CS_GPIOS_DT_SPEC_GET(DT_DRV_INST(inst))				\
```
use `SPI_CS_GPIOS_DT_SPEC_INST_GET`. 

This is a general comment really. When you have something like `DT_*(DT_DRV_INST` in 90% of cases you have some `_INST_` instance of the same macro that let you omit `DT_DRV_INST`

spectrum70 commented:
Forgot out am additional missing SPI_CS_GPIOS_DT_SPEC_GET(DT_DRV_INST(inst)) ,
changed to SPI_CS_GPIOS_DT_SPEC_INST_GET

spectrum70 commented:
I have seen that Open Alliance works reliably also with CS deasserted and asserted after each data block (OA chunk), so in case i can remove software control from the driver.

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -837,6 +981,12 @@ static int adin2111_init(const struct device *dev)
 	}
 
+	/*
+	 * Setting this for all modes, RM rev.0 Table 3, t3 >= 50ms
```
```suggestion
	 * Setting this for all modes, RM rev. A Table 3, t3 >= 50ms
```

GeorgeCGV replied:
[Publicly available RM](https://www.analog.com/media/en/technical-documentation/data-sheets/adin2111.pdf) is revision A.

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -968,21 +1137,49 @@ static const struct ethernet_api adin2111_port_api = {
+						    SPI_HOLD_ON_CS, (0)),			\
+				.slave = DT_REG_ADDR(DT_DRV_INST(inst)),			\
+				.cs = {								\
+					.gpio = COND_CODE_1(DT_INST_PROP_OR(inst, spi_oa, 0),	\
```
```suggestion
					.gpio = COND_CODE_1(DT_INST_PROP(inst, spi_oa),		\
```

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -968,21 +1137,49 @@ static const struct ethernet_api adin2111_port_api = {
+	static struct oa_tc6 name##_oa_tc6_##inst = {						\
+		.cps = 64,									\
+		.spi = &name##_config_##inst.spi,						\
+		.protected = DT_INST_PROP_OR(inst, spi_oa_protection, 0),			\
```
```suggestion
		.protected = DT_INST_PROP(inst, spi_oa_protection),				\
```

GeorgeCGV commented:
@spectrum70 regarding CS control and current approach. Won't added `SPI_LOCK_ON` to `ADIN2111_SPI_OPERATION` do the job? The SPI could be locked for the duration of OA calls. 

That way there is no need to define extra CS pin and duplicate `SPI_CONFIG_DT`. Then the `.spi` becomes:
```
		.spi = COND_CODE_1(DT_INST_PROP(inst, spi_oa),					\
			SPI_DT_SPEC_INST_GET(inst, ADIN2111_SPI_OPERATION | SPI_LOCK_ON, 0),	\
			SPI_DT_SPEC_INST_GET(inst, ADIN2111_SPI_OPERATION, 0)),			\
```

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -966,21 +1137,49 @@ static const struct ethernet_api adin2111_port_api = {
+				.operation = (ADIN2111_SPI_OPERATION) |				\
+					DT_INST_PROP(inst, duplex) |				\
+					DT_INST_PROP(inst, frame_format) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_cpol),		\
+						    SPI_MODE_CPOL, (0)) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_cpha),		\
+						    SPI_MODE_CPHA, (0)) |			\
+					COND_CODE_1(DT_INST_PROP(inst, spi_hold_cs),		\
```
The `DT_INST_PROP_OR` won't work here. The expected output is a number (_some bit_); not plain `1` or `0`. It basically follows `SPI_CONFIG_DT` from `spi.h`.

GeorgeCGV commented:
@spectrum70 considering other fixes, could you please add a minor correction to `k_thread_create` so it uses `K_KERNEL_STACK_SIZEOF` instead of `CONFIG_...`:

```
		k_thread_create(&ctx->rx_thread, ctx->rx_thread_stack,
				K_KERNEL_STACK_SIZEOF(ctx->rx_thread_stack),
				adin2111_offload_thread,
				(void *)adin, NULL, NULL,
				CONFIG_ETH_ADIN2111_IRQ_THREAD_PRIO,
				K_ESSENTIAL, K_NO_WAIT);
```

spectrum70 commented:
> @spectrum70 considering other fixes, could you please add a minor correction to `k_thread_create` so it uses `K_KERNEL_STACK_SIZEOF` instead of `CONFIG_...`:
> 
> ```
> 		k_thread_create(&ctx->rx_thread, ctx->rx_thread_stack,
> 				K_KERNEL_STACK_SIZEOF(ctx->rx_thread_stack),
> 				adin2111_offload_thread,
> 				(void *)adin, NULL, NULL,
> 				CONFIG_ETH_ADIN2111_IRQ_THREAD_PRIO,
> 				K_ESSENTIAL, K_NO_WAIT);
> ```

done

spectrum70 commented:
> @spectrum70 regarding CS control and current approach. Won't added `SPI_LOCK_ON` and `SPI_HOLD_ON_CS` (_maybe_) to `ADIN2111_SPI_OPERATION` do the job? The SPI could be locked for the duration of OA calls.
> 
> That way there is no need to define extra CS pin and duplicate `SPI_CONFIG_DT`. Then the `.spi` becomes:
> 
> ```
> 		.spi = COND_CODE_1(DT_INST_PROP(inst, spi_oa),					\
> 			SPI_DT_SPEC_INST_GET(inst, ADIN2111_SPI_OPERATION | SPI_LOCK_ON, 0),	\
> 			SPI_DT_SPEC_INST_GET(inst, ADIN2111_SPI_OPERATION, 0)),			\
> ```

It's ok, it works this way, thanks, so changed this way.

spectrum70 commented:
@GeorgeCGV there was server-side issues with automated checks yesterday, now they are all passed.

GeorgeCGV reviewed (dismissed):
@spectrum70 thank you for the fixes and OA support, LGTM.

gmarull commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -966,21 +1134,30 @@ static const struct ethernet_api adin2111_port_api = {
 		.offload_sem = Z_SEM_INITIALIZER(name##_data_##inst.offload_sem, 0, 1),         \
 		.lock = Z_MUTEX_INITIALIZER(name##_data_##inst.lock),				\
 		.buf = name##_buffer_##inst,							\
+		.oa = DT_INST_PROP_OR(inst, spi_oa, 0),						\
```
`DT_INST_PROP(inst, spi_oa)` is sufficient, since it's a boolean prop

spectrum70 commented:
Since there is still time before the next release, i added a further related patch on this branch enabling DMA, this way the ADI boards can reach 20Mbit/s of SPI transfer rate out of the box, needed for the maximum adin2111/1110 throughput.

spectrum70 commented:
unfortunately, on further testing by zperf found  some OA related issues, freezing to solve issues.

GeorgeCGV commented:
> stm32 re-transmit same tcp sequence packet

Does `eth_adin2111_oa_data_write` return any error? What is the state of `TXBOE`, `RXBOE`, and `TXBUE` in the `STATUS0` register when it happens (maybe modification of FIFO sizes could help if the host is not fast enough)?

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -149,13 +126,159 @@ int eth_adin2111_reg_read(const struct device *dev, const uint16_t reg,
+	spi_release_dt(&cfg->spi);
+
+	return ret;
+
```
```suggestion
```

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -418,6 +558,24 @@ static int adin2111_port_send(const struct device *dev, struct net_pkt *pkt)
+
+		ret = eth_adin2111_oa_data_write(cfg->adin, pkt);
+		if (ret < 0) {
+			LOG_ERR("Failed to write OA spi data, %d", ret);
```
```suggestion
			eth_stats_update_errors_tx(data->iface);
			LOG_ERR("Failed to write OA spi data, %d", ret);
```

spectrum70 commented:
@GeorgeCGV when iperf stucks, mainly, the adin2111 receiving thread is blocked.
And loks like it is blocked at 

`/* await INT */
k_sem_take(&ctx->offload_sem, K_FOREVER);`

In this state, packets can still be sent out but nothing is received, becouse there is no hw interrupt
issued from adin2111 (checked by oscilloscope).

So strangely, only for Open Alliance, fast packet transmission by iperf locks the adin2111.
Still investigating.

GeorgeCGV commented:
> Well, looks like i miss to check for space before sending a frame in OA mode.

That is great. Feel free to change the status to "Ready for review" when it works as expected.

spectrum70 commented:
Hi,
over hard zperf/iperf testing there was a couple of issues on using oa_tc6 developed for Microchip that i could not solve without putting hands in the module, that i cannot re-test for Microchip then. So i re-implemented Open Alliance part inside the driver, featuring a single dma transfer per frame, that works much better.

oa_tc6 common code cannot be used in a profitable way, due to : 
- not addressing port 2
- sends only chunk by chunk, in one dma transfer per chunk, while adin2111 allows a single dma transfer with all chiunks.

I am attaching some zperf tests.

[test-zephyr-adin2111ebz-iperf.txt](https://github.com/zephyrproject-rtos/zephyr/files/14483403/test-zephyr-adin2111ebz-iperf.txt)


Waiting for comments.

PS: there is still a brief outstanding issue open, not related to this job, but i am investigating on it. I noticed that on incoming udp packet of size 1512, generic spi fails on  "Still some length to go 2" net pkt message.

spectrum70 committed:
drivers: ethernet: adin2111: fix real rx frame size

Fix issue related to "generic SPI" mode only, when a packet of
1512 bytes is received, net_pkt_write() fails and thrwos the error:

"Still some length to go 2".

This is due to net_pkt_rx_alloc_with_buffer() allocating a maximum
mtu/size of 1514, and driver is not removing 4 bytes of crc32 from
rx buffer, that comes to be 1516 (2 bytes over buffer limit).

Fix generic SPI rx frame size removing crc32 bytes.

Signed-off-by: Angelo Dureghello <adureghello@baylibre.com>

spectrum70 commented:
Fixed the issue reported above ("Still some length to go 2" net pkt message.).
Retested all, attaching zperf test.
[test-zephyr-adin2111ebz-iperf.txt](https://github.com/zephyrproject-rtos/zephyr/files/14493924/test-zephyr-adin2111ebz-iperf.txt)

spectrum70 commented:
Adding also new zperf board to board measurement, udp is at 8.5Mbit/s
[test-zephyr-adin2111ebz-iperf.txt](https://github.com/zephyrproject-rtos/zephyr/files/14512875/test-zephyr-adin2111ebz-iperf.txt)

carlocaione reviewed (approved):
not my usual cup of tea but LGTM for the general part

GeorgeCGV reviewed (approved):
Nice work, tested generic spi with enabled crc, everything works.

GeorgeCGV commented on drivers/ethernet/eth_adin2111.c:
```diff
@@ -320,6 +637,22 @@ static void adin2111_offload_thread(void *p1, void *p2, void *p3)
 			adin2111_port_on_phyint(ctx->port[1]);
 		}
 
+		if (ctx->oa) {
+			if (status1 & ADIN2111_STATUS1_P1_RX_RDY) {
+				ret = eth_adin2111_oa_data_read(dev, 0);
+				if (ret < 0) {
+					break;
+				}
+			}
+			if (status1 & ADIN2111_STATUS1_P2_RX_RDY) {
+				ret = eth_adin2111_oa_data_read(dev, 1);
+				if (ret < 0) {
+					break;
+				}
+			}
+			goto continue_unlock;
+		}
```
@spectrum70 , completely missed during the review, there should be no `break` for the `ctx->oa` code; as there is no `while`. The `break` will terminate the irq offload thread. Because the thread is `K_ESSENTIAL` the kernel should crash.
